assets/scripts/segments/segment-dict.js

```diff
@@ -0,0 +1,233 @@
+ * component group and the segment id.
+ *
+ * @param {string} group
+ * @param {id} id
+ * @returns {object} segmentInfo
+ */
+function getSegmentInfo (group, id) {
+  return (SEGMENT_INFO[group] && SEGMENT_INFO[group][id]) || SEGMENT_UNKNOWN
+}
+
+/**
+ * Retrieves the information for all items that make up a particular component group by
+ * looking up the component group and item id in `SEGMENT_INFO`.
+ *
+ * @param {string} group - component group name (i.e. `lanes`, `objects`, or `vehicles`)
+ * @param {Array} groupItems - items that make up the component group in shape of [{ id, variants }]
+ * @returns {object} componentGroupInfo - returns object in shape of { id: { characteristics, rules, variants } }
+ */
+function getComponentGroupInfo (group, groupItems) {
+  return groupItems.reduce((obj, item) => {
+    obj[item.id] = getSegmentInfo(group, item.id)
+    return obj
+  }, {})
+}
+
+/**
+ * Retrieves all graphics definitions for the component group based on each `variant`
+ * specified for each component group item.
+ *
+ * @param {Array} groupItems - items that make up the component group in shape of [ { id, variants } ]
+ * @param {object} componentGroupInfo - definition of each item from `SEGMENT_INFO` in shape of { id: { characteristics, rules, variants } }
+ * @returns {Array} componentGroupVariants - returns array of graphic definitions in shape of [ { graphics } ]
+ */
+function getComponentGroupVariants (groupItems, componentGroupInfo) {
+  return groupItems.reduce((array, item) => {
+    const { id, variants } = item
+    // groupItemVariants - all variants possible for the particular group item
+    const groupItemVariants = componentGroupInfo[id] && componentGroupInfo[id].variants
+
+    if (groupItemVariants) {
+      Object.entries(variants).forEach(([variantName, variantKey]) => {
+        // variantInfo - graphics definition for specific variants defined by group item
+        const variantInfo = (groupItemVariants[variantName] && groupItemVariants[variantName][variantKey]) || SEGMENT_UNKNOWN_VARIANT
+        array.push(variantInfo.graphics)
+      })
+    }
+
+    return array
+  }, [])
+}
+
+/**
+ * A segment may contain multiple graphics for the same graphics type ('left', 'right', 'repeat', 'center').
+ * When encountering a new graphic for the same graphics type, this function creates a new array that
+ * contains both the new graphic as well as the existing graphics.
+ *
+ * @param {Array|string} target
+ * @param {Array|string} source
+ * @returns {Array} graphicsInfo
+ */
+function appendVariantSprites (target, source) {
+  let graphicsInfo
+
+  if (Array.isArray(target) && Array.isArray(source)) {
+    graphicsInfo = [...target, ...source]
+  } else if (Array.isArray(target)) {
+    graphicsInfo = [...target, source]
+  } else if (Array.isArray(source)) {
+    graphicsInfo = [ target, ...source ]
+  } else {
+    graphicsInfo = (target === source) ? target : [ target, source ]
+  }
+
+  return graphicsInfo
+}
```

Comment:
We might be able to make this easier to understand. (CodeClimate points out that this is too complex.)

It seems that either the `target` or `source` arguments can be either a string or an array, and we need to return an array where `target` and `source` are concatenated. This function uses an `if` statement to decide what to do in each of the four cases, where `target` and `source` are both arrays, `target` is an array but `source` is not, `source` is an array but `target` is not, and when neither `target` nor `source` or arrays. The last case has an additional case where `target` and `source` may be equal to each other.

Would there be any problems if we thought about this way:

- We know that we want an array result.
- It is easy to create a new array from two source arrays concatenated together.
- Therefore, first convert `source` or `target` to a single-length array, if they are not already arrays.
- Next, concatenate `source` to `target`.
- To prevent duplicates, we can [filter the resulting list at the end](https://lodash.com/docs#uniq)